vision_evaluation/evaluators.py

```diff
@@ -263,10 +268,79 @@ def __init__(self, prediction_filter):
     def _get_id(self):
         return f'precision_{self.prediction_filter.identifier}'
 
+    def calculate_score(self, average='macro'):
+        return super(PrecisionEvaluator, self).calculate_score(average=average, filter_out_zero_tgt=False)
+
     def _calculate(self, targets, predictions, average):
+        """
+        There is a special case that needs to be handled appropriately.
+        Due to filtering conditions, there are occasions where only 1 class remains in targets/predictions and sklearn interpretes this as an invalid configuration for multilabel.
+        Since when average == 'samples' is only calculated for multilabel, sm.recall_score throws an exception. e.g.:
+        predictions = np.array([[True]])
+        targets = np.array([[True]])
+        recall_score(targets, predictions, average='samples') throws the following error:
+        ValueError: Samplewise metrics are not available outside of multilabel classification.
+        """
+        if targets.shape[1] == 1 and average == 'samples':
+            targets = np.append(targets, np.zeros((targets.shape[0], 1)), axis=1)
+            predictions = np.append(predictions, np.zeros((predictions.shape[0], 1)), axis=1)
         return sm.precision_score(targets, predictions, average=average)
 
 
+class MeanAveragePrecisionAtK(MemorizingEverythingEvaluator):
+    """
+    MeanAveragePrecision @ K as defined here:
+    https://stackoverflow.com/questions/54966320/mapk-computation
+    https://medium.com/@misty.mok/how-mean-average-precision-at-k-map-k-can-be-more-useful-than-other-evaluation-metrics-6881e0ee21a9
+    MAP@K is the average of AveP(K) over all queries. Hence it uses average="samples".
+    AveP(K) =  Num / Den, where:
+    Num = sum_i^k P(i) * rel(i), where:
+    P(i) is the Precision @ i
+    rel(i) is an indicator function where rel(i) = 1 if position i is relevant, and rel(i) = 0 otherwise.
+    Den = min(K, number of relevant images)
+    """
+
+    def __init__(self, rank):
+        super(MeanAveragePrecisionAtK, self).__init__()
+        self.rank = rank
+
+    def _get_id(self):
+        return f'map_top{self.rank}'
+
+    def calculate_score(self, average='samples'):
+        assert average == 'samples'
+        if self.rank == 0:
+            return 0.0
+        return self._calculate(average=average)
+
+    def _calculate(self, average='samples'):
+        assert self.all_predictions.shape == self.all_targets.shape
+        if self.all_predictions.size == 0:
+            return 0.0
+        return np.mean([self._average_precision_at_k(preds, targets) for preds, targets in zip(self.all_predictions, self.all_targets)])
+
+    def _average_precision_at_k(self, predictions, targets):
+        total_pos_gt = np.sum(targets)
+        if total_pos_gt == 0:
+            return 0.0
+        rank = min(self.rank, len(predictions))
+        precision_eval = PrecisionEvaluator(TopKPredictionFilter(rank))
```

Comment:
I am confused at this part. Line 320 is calculating the AP@K for each instance, which means PrecisionEvaluator is created and calculated for each (prediction, gt) pair, and as average is set to None, it is "class-wise" calculation. Given prediction is probably never zero, it ends up to the score ends up being equal to targets, which \in [0, 1].

Reply:
you are right, it's completely redundant